### PR TITLE
refactor: `clap` upgrade to v4 

### PR DESCRIPTION
**Pull request by CHOUMnote**
Previously, we were using `version 3 of Clap`, so we upgraded to `version 4`. However, this caused compile errors due to the version upgrade, leading us to remove a lot of legacy code and refactor the project.

Note) I tested the `debug` option as thoroughly as possible, but I was unable to test the `dump` option because I didn't know how to use it.

### CODE REVIEW

File: raftify/src/cli/mod.rs [hunk 1]
```diff
@@ -15,116 +15,103 @@ use crate::{
...
+// Subcommand Select
+#[derive(Subcommand)]
+enum Commands {
+    /// Debug tools
+    #[command(subcommand)]
+    Debug(DebugSubcommands),
+    /// Dump tools
+    Dump(Dump),
+    /// plz, list the following functions
+    Nan,
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:43]: `Nan` seems to have been in the example code
Please delete it.

File: raftify/src/cli/mod.rs [hunk 2]
```diff
@@ -15,116 +15,103 @@ use crate::{
...
+    /// The peers to dump
+    peers: String,
+}
+
 pub async fn cli_handler<
     LogEntry: AbstractLogEntry + Debug + Send + 'static,
     FSM: AbstractStateMachine + Debug + Clone + Send + Sync + 'static,
 >(
-    custom_args: Option<Vec<String>>,
+    _custom_args: Option<Vec<String>>,
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:82]: The function is also used when exposing the CLI module to the Python side. When calling `cli_handler` from the Rust side, `custom_args` is None, but when calling from the Python side, you need to pass a value to `custom_args` and parse it accordingly.

So, I suggest making the following changes

```rs
pub async fn cli_handler<
    LogEntry: AbstractLogEntry + Debug + Send + 'static,
    FSM: AbstractStateMachine + Debug + Clone + Send + Sync + 'static,
>(
    args: Option<Vec<String>>,
) -> Result<()> {
    let app = match args {
        Some(args) => App::parse_from(args),
        None => App::parse(),
    };
...
```

File: raftify/src/cli/mod.rs [hunk 3]
```diff
@@ -15,116 +15,103 @@ use crate::{
...
         }
-    };
-
-    if let Some(("dump", dump_matches)) = matches.subcommand() {
-        if let Some(path) = dump_matches.value_of("path") {
-            if let Some(peers) = dump_matches.value_of("peers") {
-                dump_peers(path, parse_peers_json(peers).unwrap(), logger.clone())?;
-            }
+        _ => {
+            panic!("Command is not exist~");
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:112]: This part is unnecessary if you remove NaN from Command.

File: raftify/src/cli/mod.rs [hunk 4]
```diff
@@ -15,116 +15,103 @@ use crate::{
...
+    Persisted {
+        /// The log directory path
+        path: String,
+    },
+    /// List persisted log entries and metadata for all local nodes
+    PersistedAll {
+        /// The log directory path
+        path: String,
+    },
+    ///List all log entries
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:58]: Missing whitespace.

```
/// List all log entries
```

File: raftify/src/cli/mod.rs [hunk 5]
```diff
@@ -15,116 +15,103 @@ use crate::{
     AbstractLogEntry, AbstractStateMachine, CustomFormatter, Result,
 };
 
+/// It used nested subcommands struct, using clap 
+/// [`Commands`] -> represent functions
+/// [`DebugSubcommands`], [`Dump`] -> these parse additional factors for commands to use.
+/// if you enter the command like `raftify-cli debug persisted-all ./logs`
+/// it's expected [`App::command`] is debug, [`DebugSubcommands::PersistedAll::path`] is./logs
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:22]: This comment seems unnecessary.
Please delete

**Review by jopemachine — CHANGES_REQUESTED** [review]
Thanks for your PR!
Please check the review

**Comment by CHOUMnote** [comment]
@jopemachine I have applied the code. Please review it.

File: raftify/src/cli/mod.rs [hunk 6]
```diff
@@ -15,116 +15,98 @@ use crate::{
     AbstractLogEntry, AbstractStateMachine, CustomFormatter, Result,
 };
 
+/// CLI parser
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:18]: Please do not leave unnecessary comments.

File: raftify/src/cli/mod.rs [hunk 7]
```diff
@@ -15,116 +15,98 @@ use crate::{
...
+#[command(version = PKG_VERSION)]
+#[command(author = PKG_AUTHORS)]
+#[command(about = PKG_DESCRIPTION)]
+struct App {
+    /// required main commands
+    #[command(subcommand)]
+    command: Commands,
+}
+
+// Subcommand Select
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:30]: Same here

File: raftify/src/cli/mod.rs [hunk 8]
```diff
@@ -15,116 +15,98 @@ use crate::{
...
 };
 
+/// CLI parser 
+#[derive(Parser)]
+#[command(name = "raftify")]
+#[command(version = PKG_VERSION)]
+#[command(author = PKG_AUTHORS)]
+#[command(about = PKG_DESCRIPTION)]
+struct App {
+    /// required main commands
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:25]: Same here

File: raftify/src/cli/mod.rs [hunk 9]
```diff
@@ -15,116 +15,98 @@ use crate::{
...
-                        .index(2),
-                )
-                .about(""),
-        );
-
-    let matches = match custom_args {
-        Some(args) => app.get_matches_from(args),
-        None => app.get_matches(),
+    let app: App = match args{
+        // using python side
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:79]: Same here

File: raftify/src/cli/mod.rs [hunk 10]
```diff
@@ -15,116 +15,98 @@ use crate::{
...
+        Some(args) => App::parse_from(args),
+        // using env::args
```
Comment (jopemachine) [inline raftify/src/cli/mod.rs:81]: There are several unnecessary whitespaces here.
Running `cargo fmt` before committing can prevent such mistakes.

**Comment by CHOUMnote** [comment]
i have a question. do you recommend not to write comments as much as possible?

**Comment by jopemachine** [comment]
> i have a question. do you recommend not to write comments as much as possible?

No, I'm not saying don't comment; rather, the code should be self-explanatory.

Comments should explain things that cannot be explained by the code itself, not repeat what the code already expresses.

I think this code is already sufficiently self-explanatory, so additional comments are not necessary.

**Comment by jopemachine** [comment]
@CHOUMnote Thanks for your contribution! ❤️

**Pull request merged**